concatenated AWS policy statements fail with data type error

**Describe the issue**

When scanning an aws_s3_bucket_policy where the statements are the result of a "concat" operation, checkov errors with the following:

```
TypeError: string indices must be integers, not 'str'
```

**Examples**

The culprit resource is:

```
resource "aws_s3_bucket_policy" "logs" {
  bucket = aws_s3_bucket.logs.id
  policy = jsonencode({
    Version = "2012-10-17"
    Statement = concat(
      jsondecode(data.aws_iam_policy_document.logs-cloudtrail-policy-acl-check.json).Statement,
      jsondecode(data.aws_iam_policy_document.s3-logs-cloudtrail-policy-write.json).Statement,
      jsondecode(data.aws_iam_policy_document.s3-logs-vpc-flow-logs-policy.json).Statement,
    )
  })
}
```

Each referenced aws_iam_policy_document is valid in of itself.  The problem appears to be that the statement is not evaluated to a real dict value before checkov attempts to introspect it.  Instead, checkov is looking at the string `concat(....)`, not a data structure.


**Exception Trace**

```
2025-04-08 12:52:24,558 [ThreadPoolEx] [ERROR]  Failed to run check CKV_AWS_70 on /s3.tf:aws_s3_bucket_policy.homeboost_logs
Traceback (most recent call last):
  File "/Users/myddryn/venv/localpy/lib/python3.11/site-packages/checkov/common/checks/base_check.py", line 68, in run
    check_result["result"] = self.scan_entity_conf(entity_configuration, entity_type)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/myddryn/venv/localpy/lib/python3.11/site-packages/checkov/terraform/checks/resource/base_resource_check.py", line 43, in scan_entity_conf
    return self.scan_resource_conf(conf)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/myddryn/venv/localpy/lib/python3.11/site-packages/checkov/terraform/checks/resource/aws/S3AllowsAnyPrincipal.py", line 86, in scan_resource_conf
    if statement['Effect'] == 'Deny' or 'Principal' not in statement:
       ~~~~~~~~~^^^^^^^^^^
TypeError: string indices must be integers, not 'str'
```

**Desktop (please complete the following information):**
 - OS: Mac OS, 15.3.2
 - Checkov Version: 3.2.400

--- a/checkov/arm/checks/resource/StorageAccountName.py
+++ b/checkov/arm/checks/resource/StorageAccountName.py
@@ -9,7 +9,7 @@ from checkov.arm.base_resource_check import BaseResourceCheck
 
 STO_NAME_REGEX = re.compile(r"^[a-z0-9]{3,24}$")
 VARIABLE_REFS = ("local.", "module.", "var.", "random_string.", "random_id.", "random_integer.", "random_pet.",
-                 "azurecaf_name", "each.")
+                 "azurecaf_name", "each.", "substring")
 
 
 class StorageAccountName(BaseResourceCheck):

--- a/checkov/terraform/checks/graph_checks/aws/Route53ZoneEnableDNSSECSigning.yaml
+++ b/checkov/terraform/checks/graph_checks/aws/Route53ZoneEnableDNSSECSigning.yaml
@@ -4,11 +4,18 @@ metadata:
   category: "NETWORKING"
 definition:
   or:
-    - cond_type: attribute
-      resource_types:
-        - aws_route53_zone
-      attribute: vpc  # This indicates a private zone that can't have DNSSEC enabled
-      operator: exists
+    - and:
+      - cond_type: attribute
+        resource_types:
+          - aws_route53_zone
+        attribute: vpc  # This indicates a private zone that can't have DNSSEC enabled
+        operator: exists
+      - cond_type: attribute
+        resource_types:
+          - aws_route53_zone
+        attribute: vpc  # This indicates a private zone that can't have DNSSEC enabled
+        operator: not_equals
+        value: []
     - and:
       - cond_type: filter
         attribute: resource_type

--- a/checkov/terraform/checks/graph_checks/aws/Route53ZoneEnableDNSSECSigning.yaml
+++ b/checkov/terraform/checks/graph_checks/aws/Route53ZoneEnableDNSSECSigning.yaml
@@ -3,28 +3,23 @@ metadata:
   id: "CKV2_AWS_38"
   category: "NETWORKING"
 definition:
- and :
-    - cond_type: filter
-      attribute: resource_type
-      value:
-      - aws_route53_zone
-      operator: within
-    - cond_type: connection
+  or:
+    - cond_type: attribute
       resource_types:
-      - aws_route53_zone
-      connected_resource_types:
-      - aws_route53_hosted_zone_dnssec
-      - aws_route53_key_signing_key
+        - aws_route53_zone
+      attribute: vpc  # This indicates a private zone that can't have DNSSEC enabled
       operator: exists
-    - or:
-      - cond_type: attribute
-        resource_types:
-          - aws_route53_zone
-        attribute: vpc
-        operator: not_exists
+    - and:
+      - cond_type: filter
+        attribute: resource_type
+        value:
+        - aws_route53_zone
+        operator: within
       - cond_type: connection
         resource_types:
         - aws_route53_zone
         connected_resource_types:
-        - aws_route53_zone_association
-        operator: not_exists
+        - aws_route53_hosted_zone_dnssec
+        - aws_route53_key_signing_key
+        - aws_route53_zone_association  # This indicates a private zone that can't have DNSSEC enabled
+        operator: exists

--- a/checkov/terraform/checks/resource/aws/S3AllowsAnyPrincipal.py
+++ b/checkov/terraform/checks/resource/aws/S3AllowsAnyPrincipal.py
@@ -82,8 +82,11 @@ class S3AllowsAnyPrincipal(BaseResourceCheck):
                     return CheckResult.UNKNOWN
 
         if isinstance(policy_block, dict) and 'Statement' in policy_block.keys():
-            for statement in force_list(policy_block['Statement']):
-                if statement['Effect'] == 'Deny' or 'Principal' not in statement:
+            statements = force_list(policy_block['Statement'])
+            if all('Effect' not in statement for statement in statements):
+                return CheckResult.UNKNOWN
+            for statement in statements:
+                if 'Effect' not in statement or statement['Effect'] == 'Deny' or 'Principal' not in statement:
                     continue
                 principal = statement['Principal']
                 if principal == '*':

--- a/tests/arm/checks/resource/example_StorageAccountName/substring.bicep
+++ b/tests/arm/checks/resource/example_StorageAccountName/substring.bicep
@@ -1,0 +1,52 @@
+@description('Name of the environment')
+param environmentName string
+
+@description('Name of the Storage account')
+param storageAccountName string = substring('abcdefgh${environmentName}${uniqueString(resourceGroup().id)}', 0, 24)
+
+@description('Provide a location for the resources.')
+param location string = resourceGroup().location
+
+resource dataStorageAccount 'Microsoft.Storage/storageAccounts@2021-09-01' = {
+  name: storageAccountName
+  location: location
+  sku: {
+    name: 'Standard_LRS'
+  }
+  kind: 'StorageV2'
+  identity: {
+    type: 'SystemAssigned'
+  }
+  properties: {
+    accessTier: 'Hot'
+    allowBlobPublicAccess: false
+    allowSharedKeyAccess: true
+    allowCrossTenantReplication: false
+    isHnsEnabled: true
+    allowedCopyScope: 'AAD'
+    defaultToOAuthAuthentication: false
+    encryption: {
+      keySource: 'Microsoft.Storage'
+      requireInfrastructureEncryption: false
+      services: {
+        blob: {
+          enabled: true
+          keyType: 'Account'
+        }
+      }
+    }
+    minimumTlsVersion: 'TLS1_2'
+    largeFileSharesState: 'Disabled'
+    sasPolicy: {
+      expirationAction: 'Log'
+      sasExpirationPeriod: '00.00:10:00'
+    }
+    supportsHttpsTrafficOnly: true
+    networkAcls: {
+      bypass: 'AzureServices'
+      virtualNetworkRules: []
+      ipRules: []
+      defaultAction: 'Allow'
+    }
+  }
+}

--- a/tests/terraform/checks/resource/aws/example_S3AllowsAnyPrincipal/main.tf
+++ b/tests/terraform/checks/resource/aws/example_S3AllowsAnyPrincipal/main.tf
@@ -484,3 +484,16 @@ resource "aws_s3_bucket" "pass_w_condition6" {
 }
 POLICY
 }
+
+# Handle error
+resource "aws_s3_bucket_policy" "logs" {
+  bucket = aws_s3_bucket.logs.id
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = concat(
+      jsondecode(data.aws_iam_policy_document.logs-cloudtrail-policy-acl-check.json).Statement,
+      jsondecode(data.aws_iam_policy_document.s3-logs-cloudtrail-policy-write.json).Statement,
+      jsondecode(data.aws_iam_policy_document.s3-logs-vpc-flow-logs-policy.json).Statement,
+    )
+  })
+}

--- a/tests/terraform/graph/checks/resources/Route53ZoneEnableDNSSECSigning/expected.yaml
+++ b/tests/terraform/graph/checks/resources/Route53ZoneEnableDNSSECSigning/expected.yaml
@@ -2,5 +2,6 @@ pass:
   - "aws_route53_zone.pass"
   - "aws_route53_zone.private_with_inline_vpc"
   - "aws_route53_zone.private_with_zone_association"
+  - "aws_route53_zone.pass_signing_key"
 fail:
-  - "aws_route53_zone.fail"
+  - "aws_route53_zone.fail2"

--- a/tests/terraform/graph/checks/resources/Route53ZoneEnableDNSSECSigning/main.tf
+++ b/tests/terraform/graph/checks/resources/Route53ZoneEnableDNSSECSigning/main.tf
@@ -34,13 +34,14 @@ resource "aws_route53_zone_association" "private" {
   vpc_id  = "vpc-1a2b3c4d"
 }
 
-#fail
-resource "aws_route53_zone" "fail" {
-  name = "fail"
+#pass with signing key
+resource "aws_route53_zone" "pass_signing_key" {
+  name = "pass"
 }
+
 resource "aws_route53_key_signing_key" "fail" {
-  hosted_zone_id             = aws_route53_zone.fail.id
-  key_management_service_arn = aws_kms_key.fail.arn
+  hosted_zone_id             = aws_route53_zone.pass_signing_key.id
+  key_management_service_arn = aws_kms_key.pass_signing_key.arn
   name                       = "pass"
 }
 

--- a/tests/terraform/graph/checks/test_yaml_policies.py
+++ b/tests/terraform/graph/checks/test_yaml_policies.py
@@ -568,6 +568,9 @@ class TestYamlPolicies(unittest.TestCase):
     def test_Route53ZoneHasMatchingQueryLog(self):
         self.go("Route53ZoneHasMatchingQueryLog")
 
+    def test_Route53ZoneEnableDNSSECSigning(self):
+        self.go("Route53ZoneEnableDNSSECSigning")
+
 
     def test_registry_load(self):
         registry = Registry(parser=GraphCheckParser(), checks_dir=str(

